### PR TITLE
feat(ssa): Hoist MakeArray instructions during loop invariant code motion 

### LINKED ISSUE #6775: Instantiate global arrays only once
### Aim

I should be able to use large global array constants in a loop. Global constants cannot be mutable so I should be able to just be performing a read on an already existing array value. 

### Expected Behavior

The loop from `regression_4709` should not take very long to compute:
```
    let mut acc: Field = 0;
    for i in 0..257 {
        for j in 0..257 {
            acc += EXPONENTIATE[i][j];
        }
    }
    assert(!acc.lt(x));
    assert(x != y);
```
However, `regression_4709` is not run as part of our brillig execution opcodes report as it takes too much time/memory. 

### Bug

Reading from a global array in a loop can lead to a massive execution blowup as we are generating the array and incrementing its reference counter inside of the loop.

If we take `regression_4709`, and reduce the `EXPONENTIATE` array to be 10 x 10 rather than 257 x 257 we can see this blowup. 
```
global EXPONENTIATE: [[Field; 10]; 10] = [
    [1, 1, 1, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
];
fn main(x: Field, y: pub Field) {
    let mut acc: Field = 0;
    for i in 0..10 {
        for j in 0..10 {
            acc += EXPONENTIATE[i][j];
        }
    }
    assert(!acc.lt(x));
    assert(x != y);
}
```
The initial SSA we generate is the following:
```
brillig(inline) fn main f0 {
  b0(v0: Field, v1: Field):
    v4 = allocate -> &mut Field
    store Field 0 at v4
    jmp b1(u32 0)
  b1(v2: u32):
    v8 = lt v2, u32 10
    jmpif v8 then: b3, else: b2
  b2():
    v9 = load v4 -> Field
    v11 = call f1(v9, v0) -> u1
    v12 = not v11
    constrain v11 == u1 0
    v14 = eq v0, v1
    v15 = not v14
    constrain v14 == u1 0
    return
  b3():
    jmp b4(u32 0)
  b4(v3: u32):
    v16 = lt v3, u32 10
    jmpif v16 then: b6, else: b5
  b5():
    v18 = add v2, u32 1
    jmp b1(v18)
  b6():
    v19 = load v4 -> Field
    v21 = make_array [Field 1, Field 1, Field 1, Field 1, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v22 = make_array [Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v23 = make_array [Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v24 = make_array [Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v25 = make_array [Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v26 = make_array [Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v27 = make_array [Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v28 = make_array [Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v29 = make_array [Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v30 = make_array [Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v31 = make_array [v21, v22, v23, v24, v25, v26, v27, v28, v29, v30] : [[Field; 10]; 10]
    v32 = array_get v31, index v2 -> [Field; 10]
    inc_rc v32
    v33 = array_get v32, index v3 -> Field
    v34 = add v19, v33
    store v34 at v4
    v35 = add v3, u32 1
    jmp b4(v35)
}
```
You will see that we are creating our nested array inside of the inner most nested loop. If we were to switch the global to a let statement inside of `main` we see these `make_array` instructions are laid down in `b0`. The global constant version executes 12.5k brillig opcodes and the local variable version executes 2.2k.

Once we get to constant folding the repeat `make_array` instructions are replaced by `inc_rc` instructions due to this logic when deduplicating make arrays https://github.com/noir-lang/noir/blob/f6957faf50025afc053a32d86398d0823253066b/compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs#L321.

Our final SSA then looks like this:
```
brillig(inline) fn main f0 {
  b0(v0: Field, v1: Field):
    v4 = allocate -> &mut Field
    store Field 0 at v4
    jmp b1(u32 0)
  b1(v2: u32):
    v8 = lt v2, u32 10
    jmpif v8 then: b3, else: b2
  b2():
    v9 = load v4 -> Field
    v11 = call field_less_than(v9, v0) -> u1
    constrain v11 == u1 0
    v13 = eq v0, v1
    constrain v13 == u1 0
    return
  b3():
    jmp b4(u32 0)
  b4(v3: u32):
    v14 = lt v3, u32 10
    jmpif v14 then: b6, else: b5
  b5():
    v16 = add v2, u32 1
    jmp b1(v16)
  b6():
    v17 = load v4 -> Field
    v19 = make_array [Field 1, Field 1, Field 1, Field 1, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    v20 = make_array [Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0, Field 0] : [Field; 10]
    inc_rc v20
    inc_rc v20
    inc_rc v20
    inc_rc v20
    inc_rc v20
    inc_rc v20
    inc_rc v20
    inc_rc v20
    v21 = make_array [v19, v20, v20, v20, v20, v20, v20, v20, v20, v20] : [[Field; 10]; 10]
    v22 = array_get v21, index v2 -> [Field; 10]
    inc_rc v22
    v23 = array_get v22, index v3 -> Field
    v24 = add v17, v23
    store v24 at v4
    v25 = add v3, u32 1
    jmp b4(v25)
}
```

This array creation logic should be outside of the loop and we should not see a difference in execution trace when working with a global or local array.

### To Reproduce

1. Go to `execution_success/regression_4709`
2. Reduce size of exponentiation
3. Compile with `--show-ssa`
4. Inspect SSA

### Workaround

Yes

### Workaround Description

Use a local array to reduce execution trace

### Additional Context

_No response_

### Project Impact

Blocker

### Blocker Context

There is a workaround, but unless a user is aware of this internal compiler bug they will not be aware of why their program is so slow. Thus, I have marked it as a blocker.

### Nargo Version

nargo version = 1.0.0-beta.0 noirc version = 1.0.0-beta.0+f1756ef1b2fe78da24d250bba3534870a00685d4 (git version hash: 2a840ca5fe1e37e057f3fb0b301da205dbbe6080, is dirty: true)

### NoirJS Version

_No response_

### Proving Backend Tooling & Version

_No response_

### Would you like to submit a PR for this Issue?

Yes

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves #6775 

## Summary\*

In https://github.com/noir-lang/noir/pull/6685 we restricted that `MakeArray` can only be deduplicated in an ACIR runtime. This is due to that arrays can actually be mutated in unconstrained code. 

Now when hoisting loop invariants, we check whether we are hoisting a `MakeArray`. If we have hoisted a `MakeArray` instruction, we also insert an `IncrementRc` instruction in the old location of `MakeArray` in the loop block.

This PR is trading off the cost of creating an array inside of a loop vs. the cost of creating an array once + incrementing its reference counter inside the loop. Other than for the smallest of arrays, this tradeoff should almost always be worth it. 

Before merging we may want to determine some kind of heuristic for when this optimization should run. It could also be left for a follow-up PR. 

## Additional Context

A potential optimization to follow-up this PR would be to bring back the rc tracker that was fully removed in DIE https://github.com/noir-lang/noir/pull/6700. This would more generally help with the cost from deduplicating `MakeArray` instructions more generally as right now we always insert an `IncrementRc` upon deduplicating or hoisting a `MakeArray`. If that array is never mutably borrowed in the block, DIE should be able to remove those instructions that were included to maintain correctness. 

`regression_4709` performance:
Master:
The test was previously excluded from our CI's Brillig tests due to that execution would hang. I even ran out of memory and crashed the mainframe when trying to profile it.
After this PR:
~52 million opcodes executed
After this PR w/ https://github.com/noir-lang/noir/pull/6783:
~1.2 millions opcodes executed

`blob` performance:
Master:
~990 million opcodes executed
After this PR:
~775 million opcodes executed (23.7% improvement from master)
After this PR w/ https://github.com/noir-lang/noir/pull/6783:
~449 million opcodes executed (54.6% improvement from master)

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs [hunk 1]
```diff
@@ -113,6 +114,29 @@ impl<'f> LoopInvariantContext<'f> {
...
+                    ) {
+                        let results =
+                            self.inserter.function.dfg.instruction_results(instruction_id).to_vec();
+                        let results = vecmap(results, |value| self.inserter.resolve(value));
+                        assert_eq!(
+                            results.len(),
+                            1,
+                            "ICE: We expect only a single result from an `Instruction::MakeArray`"
+                        );
+                        dbg!(results.clone());
```
Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs:132]:
```suggestion

```

File: test_programs/gates_report_brillig_execution.sh [hunk 2]
```diff
@@ -8,8 +8,6 @@ excluded_dirs=(
   "double_verify_nested_proof" 
   "overlapping_dep_and_mod" 
   "comptime_println"
-  #  Takes a very long time to execute as large loops do not get simplified.
-  "regression_4709"
```
Comment (vezenovm) [inline test_programs/gates_report_brillig_execution.sh:12]: This test can now be profiled in a reasonable amount of time.

**Comment by jfecher** [comment]
> This PR is trading off the cost of creating an array inside of a loop vs. the cost of creating an array once + incrementing its reference counter inside the loop. Other than for the smallest of arrays, this tradeoff should almost always be worth it.

I'll note that if we're creating an array in the loop and mutating it each iteration then this approach would be slower since we'd have an extra inc_rc instruction each iteration and need to do a dynamic check to copy the array. Overall though it's probably a fine tradeoff since the good case of this opt is much better than the worst case applying it can have.

**Comment by vezenovm** [comment]
> I'll note that if we're creating an array in the loop and mutating it each iteration

Yeah good note. The trade-off noted in the description is more for creating immutable arrays outside the loop. However, an array declared mutable inside the loop should already be issuing an inc_rc instruction after being created and will be copied already every iteration.

File: compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs [hunk 3]
```diff
@@ -109,6 +110,28 @@ impl<'f> LoopInvariantContext<'f> {
...
+                        Instruction::MakeArray { .. }
+                    ) {
+                        let results =
+                            self.inserter.function.dfg.instruction_results(instruction_id).to_vec();
+                        let results = vecmap(results, |value| self.inserter.resolve(value));
+                        assert_eq!(
+                            results.len(),
+                            1,
+                            "ICE: We expect only a single result from an `Instruction::MakeArray`"
+                        );
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs:127]: If we only expect a single result, why allocate the temporary Vec?
Also, can't we get two results in the case of making slices?

File: compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs [hunk 4]
```diff
@@ -109,6 +110,28 @@ impl<'f> LoopInvariantContext<'f> {
 
                 if hoist_invariant {
                     self.inserter.push_instruction(instruction_id, pre_header);
+
+                    // We track whether we may mutate MakeArray instructions before we deduplicate
+                    // them but we still need to issue an extra inc_rc in case they're mutated afterward.
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs:115]: > We track whether we may mutate MakeArray instructions before we deduplicate them

Where do you do this? Looks like this comment was just copy pasted from constant folding

File: compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs (hunk 3 above)
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs:127]: Slices are constructed using two instructions. One for the array representing the slice and one for the dynamic length.
e.g.
```
v3 = make_array [] : [Field]
    v4 = allocate -> &mut u32
    store u32 0 at v4
```

Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs:127]: Switched to just access the first element

File: compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs (hunk 4 above)
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs:115]: Ah yeah, I did not read the comment carefully enough. I have updated it to just say we must issue an inc_rc instruction.

**Review by jfecher — APPROVED** [review]
I think hoisting without the mutable checks should be fine. Unlike deduplication, there should be no chance the array in question is used between the hoist point and original point since the instruction just wasn't present there before.

**Pull request merged**